### DISCUSSION TITLE
Raspberry PI 4 - Kubernetes integration

### DISCUSSION THREAD

**Discussion by dhenry123** [openobserve/openobserve#944] [General]
Hi, I've just finished installing OpenObserve in my Kubernetes cluster, powered by 3 Raspberry PI4s.

I had to manage several OOMKilled during the build (docker build and on premise). I ended up removing the web/dist embed code to disassociate it from the executable.

I now have two containers, an api server and another with just the embedded application, served by "python3 -m http.server"... I had to modify (2 lines of code) the application to send xhr requests to the api server.

If you'd like to discuss this .... is here.
+++

**Comment by prabhatsharma** [comment]
@dhenry123 Yeah, we know the need for high memory during UI build, or you get OOM. Check our build files https://github.com/openobserve/openobserve/blob/faad9c964f799c0de15991de017ec6f59d79216b/deploy/build/Dockerfile#L7

This is how we tackled it.
```
RUN NODE_OPTIONS="--max-old-space-size=8192" npm run build
```

I believe you fell into this trap. Would love to know, how you solved it.

**Reply by dhenry123** [reply]
This part (cli) was built on X86_64, only the rust code was built on premise : 

- my devices (Raspberry PI 4/8Go RAM) use 64 bits pure Debian (https://raspi.debian.net/)
- clone the project
- rust install : `curl --proto '=https' --tlsv1.2 https://sh.rustup.rs -sSf | sh`
- and some stuff : 
```
apt-get install -y protobuf-compiler build-essential
```
- Debian Protobuf is not the last release, protoc must be executed with "--experimental_allow_proto3_optional" parameter, to deal with proto3. Didn't know how to deal with rust config building, so i used this hack (https://github.com/signalapp/libsignal/issues/523#issuecomment-1562970981) : 
```
mv /usr/bin/protoc /usr/bin/protoc-bin
echo "#!/bin/bash" > /usr/bin/protoc
echo "/usr/bin/protoc-bin \"\$@\" --experimental_allow_proto3_optional" >> /usr/bin/protoc
```
- then compile : 
```
# Delete cli build content
if [ -f web/dist ];then
    # compliance but empty
    rm -rf web/dist
    mkdir web/dist
fi
jobs=1
source "$HOME/.cargo/env"
RUSTFLAGS='-C target-feature=+crc,+neon' cargo build --jobs=$jobs --release --features mimalloc
```
And wait for more than 1 hour...

- Create Dockerfile, which is the last section of deploy/build/Dockerfile : 
`vi deploy/build/Dockerfile.raspi4` :
```
FROM debian:bullseye-slim as runtime
RUN apt-get update && apt-get install -y curl htop iftop sysstat procps lsof net-tools
COPY target/release/openobserve /
CMD ["/openobserve"]
```
- Build the new arm64 image 
```
docker build -f deploy/build/Dockerfile.raspi4 -t openobserve:apiv0.4.7 .
```
## Cli

**For cli, i ran into developper mode on X86_64**

set the ".env" file with VITE_OPENOBSERVE_ENDPOINT=[url of my service in kubernetes] trying to get a quick solution to get a standalone cli which connect to API installed and running in the Raspberry Kubernetes cluster.

I changed, to manage assets and the initial state : 

### web/src/utils/zincutils.ts :

```diff
--- a/web/src/utils/zincutils.ts
+++ b/web/src/utils/zincutils.ts
@@ -15,6 +15,7 @@
 import config from "../aws-exports";
 import { ref, onMounted, onUnmounted } from "vue";
 import { Dialog } from "quasar";
+import store from "../stores";
 
 const useLocalStorage = (
   key: string,
@@ -254,12 +255,13 @@ export const getImageURL = (image_path: string) => {
 
 export const getPath = () => {
   const pos = window.location.pathname.indexOf("/web/");
-  const path =
-    window.location.origin == "http://localhost:8081"
-      ? "/"
-      : pos > -1
-      ? window.location.pathname.slice(0, pos + 5)
-      : window.location.pathname;
+  const path = store.state.API_ENDPOINT
+    ? ""
+    : window.location.origin == "http://localhost:8081"
+    ? "/"
+    : pos > -1
+    ? window.location.pathname.slice(0, pos + 5)
+    : window.location.pathname;
   const cloudPath = import.meta.env.BASE_URL;
   return config.isCloud == "true" ? cloudPath : path;
 };
```

###  web/src/stores/index.ts

```diff
index c8cbbae..6de50c7 100644
--- a/web/src/stores/index.ts
+++ b/web/src/stores/index.ts
@@ -129,7 +129,7 @@ export default createStore({
       context.commit("logout");
     },
     endpoint(context, payload) {
-      context.commit("endpoint", payload);
+      if (!API_ENDPOINT) context.commit("endpoint", payload);
     },
     setIndexData(context, payload) {
       context.commit("setIndexData", payload);
```

in dev mode, connection with api (running in my kubernetes cluster, **not the localhost**)  : OK, but with "build" version no deal with the setting "VITE_OPENOBSERVE_ENDPOINT", and i don't know VueJS. 

So just manually make a patch, before building :
```
baseUrl="https://myopenobserver.service"
file="web/src/services/http.ts"
sed -i "s|baseURL: .*$|baseURL: \"$baseUrl\",|" $file
```
then build :  `NODE_OPTIONS="--max-old-space-size=8192" npm run build `

So i got a "build" with the hard coded xhr host, but expected, then build a docker image with a slim web server, and the cli build content :

add file deploy/build/Dockerfile.clionly : 

```
FROM python:alpine3.18 as runtime
COPY ./web/dist /app
WORKDIR /app
CMD ["python3", "-m", "http.server"] 
```
As i am running dev on X86_64 (with buildx) : 

`docker build --platform=arm64 -f deploy/build/Dockerfile.clionly -t openobserve:cliv0.4.7 .`

And push image in my arm64 registry

**WARN** this cli version, doesn't support dynamic route, once you use the cli with browser, url change with action, but if you press 'F5', the webserver doesn't know how to deal with route, it knows only the "/". To recall the app, use only "http(s)://[domain]/"
It's possible with apache (Alias or Rewrite), "route" with nodejs. etc....The next

I have to find the right setting to compile with cross compile for CORTEX-A72 from X86_64, i will be faster :) 

Running since 18 hours, ~40k documents ingested, Memory max 150 MO, et CPU max : 64milli
That's all :)

**Reply by dhenry123** [reply]
The size of executable is 66 Mo in my case

**Comment by dhenry123** [comment]
finally, i succeded building one binary with api and cli with cross compilation, see howto here : https://codeberg.org/mytinydc/Openobserve-Raspberry-PI-4-aarch64

Thanks for your job

**Comment by hengfeiyang** [comment]
Try this: https://github.com/openobserve/openobserve/issues/3910
